pkg/proxy/proxy.go

```diff
@@ -215,7 +214,18 @@ retryCreatePort:
 			redir.implementation, err = createKafkaRedirect(redir, kafkaConfiguration{}, DefaultEndpointInfoRegistry)
 
 		case policy.ParserTypeHTTP:
-			redir.implementation, err = createEnvoyRedirect(redir, p.stateDir, p.XDSServer, wg)
+			redir.implementation, err = createEnvoyRedirect(redir, p.stateDir, p.XDSServer, wg, func() (uint16, error) {
+				// Reallocate proxyport after a NACK has been received.
+				p.mutex.Lock()
+				newPort, err := p.allocatePort()
+				if err == nil {
+					delete(p.allocatedPorts, redir.ProxyPort)
+					redir.ProxyPort = newPort
```

Comment:
You're not holding any mutex when modifying `redir`? That's probably not thread-safe. There's a good chance this update won't be seen by other goroutines.

Reply:
Added proper locking as a separate commit so that it is easier to review.